File: 01/005/main.py
def allReducable(num, reducer):
    for i in range(1, reducer + 1):
        if num % i != 0:
            return False
    return True

def reduceResult(num, reducer):
    for i in range(reducer, 1, -1):
        if allReducable(num/i, reducer):
            return reduceResult(num/i, reducer)
    return num

def smallestDivisible(reducer):
    result = 1
    for i in range(2, reducer):
        result = reduceResult(result*i, i)
    return int(result)

print(smallestDivisible(20))

File: 01/005/main.ts
function allReducable(num: number, reducer: number) {
  for (let i = 1; i <= reducer; i++) {
    if (num % i !== 0) {
      return false;
    }
  }
  return true;
}

function reduceResult(num: number, reducer: number) {
  for (let i = reducer; i > 1; i--) {
    if (allReducable(num / i, reducer)) {
      return reduceResult(num / i, reducer);
    }
  }
  return num;
}

function smallestDivisible(reducer: number) {
  let result = 1;
  for (let i = 2; i <= reducer; i++) {
    result = reduceResult(result * i, i);
  }
  return result;
}

console.log(smallestDivisible(20));
